### PR TITLE
Back Link Service

### PR DESCRIPTION
**Pull request by ChazUK**
Adding a Back Link Service. You can set the back link by calling

```this.backService.back = { url: ['/path'] }```

It can also take `fragments` and `queryParams` as optional properties

### CODE REVIEW

File: src/app/core/services/back.service.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import { NavigationEnd, Router } from '@angular/router';
+import { URLStructure } from '@core/model/url.model';
+import { BehaviorSubject, Observable } from 'rxjs';
+import { filter, map } from 'rxjs/operators';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class BackService {
+  private _back$: BehaviorSubject<URLStructure> = new BehaviorSubject<URLStructure>(null);
```
Comment (omarCreativeDev) [inline src/app/core/services/back.service.ts]: is there any benefit to prepending an underscore? the private type should be enough no?

File: src/app/core/services/back.service.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  public back$: Observable<URLStructure> = this._back$.asObservable();
```
Comment (omarCreativeDev) [inline src/app/core/services/back.service.ts]: i dont see why this is needed if we have the above BehaviourSubject?

File: src/app/core/services/back.service.ts (hunk 1 above)
Comment (ChazUK) [reply src/app/core/services/back.service.ts]: I used the underscore for Behaviour Subjects and then without for the actual Observable that Components can subscribe to.

File: src/app/core/services/back.service.ts (hunk 2 above)
Comment (ChazUK) [reply src/app/core/services/back.service.ts]: No idea tbh, I saw this pattern being used somewhere else, I think the general premise is that it stops the direct access to the BehaviourSubject

File: src/app/core/services/back.service.ts (hunk 1 above)
Comment (omarCreativeDev) [reply src/app/core/services/back.service.ts]: I just feel as a behaviour subject is essentially an observable there's no point in creating another

File: src/app/core/services/back.service.ts (hunk 2 above)
Comment (omarCreativeDev) [reply src/app/core/services/back.service.ts]: thats fair and i see your point. however i think in our application [from what i have seen so far] the state managed observables need to be accessible by different routes/components and therefore the observable stream should be accessable/public anyway : )

**Comment by ChazUK** [comment]
Updated to better handle state shown here https://dev.to/avatsaev/simple-state-management-in-angular-with-only-services-and-rxjs-41p8

This will become redundant once NGRX or NGXS is implemented.

**Comment by ChazUK** [comment]
[Back Link Service](https://trello.com/c/CwPBM3xI/506-back-link-service)

**Pull request merged**